### ISSUE TITLE
Provide a way to keep existing markup as is (Custom Elements, Hydration)

### ISSUE THREAD

**Issue by naltatis** [jorgebucaran/hyperapp#757]
I'm looking for a way to make hyperapp respect existing markup thats inside the app and leave it untouched. My problem is quite similar to the issues @frenzzy and @gyopiazza already have pointed out (#741, #688).

Parts of my application are made up of sub-applications provided through Custom Elements. Due to browser support be can not rely on ShadowDOM, so these Custom Elements use the Light DOM to render their UI. Hyperapp is overwriting their content in every render cycle. I've tried the [hydration workaround](https://github.com/hyperapp/render/issues/8#issuecomment-385974406). But this is not really a solution. You are able to restore the markup, but the innerHTML gets completely rewritten so you loose all events.

I've compiled a little JSBin the shows one concrete scenario I'm having problems with:
https://jsbin.com/dogujuq/2/edit?html,js,output

Using react I was able to make this work using `shouldComponentUpdate`. But I currently don't see any solution how to make this happen in hyperapp.

**Comment by jorgebucaran** [comment]
@naltatis What about using Hyperapp to create your custom element instead of setting innerHTML. I did [something similar](https://codepen.io/anon/pen/bjwaKz?editors=0010) with Superfine the other day when I was toying with custom elements.

**Comment by naltatis** [comment]
@jorgebucaran We use custom elements as an abstraction between teams, so could be, that a team also uses hyperapp to build their feature, but when composing features through hyperapp I can't make any assumptions about the inner workings of the sub applications.

**Comment by jorgebucaran** [comment]
@naltatis I see. I never use innerHTML like that (or at all), so I'm having a hard time understanding what's the problem (or with the issues linked above).

**Comment by frenzzy** [comment]
Looks like this case is different from innerHTML attribute hydration/recycling and probably simpler to implement. 

Instead of adding a special `shouldComponentUpdate` property to VNode Hyperapp may detect custom element and always do nothing with child nodes in this case.

But I'm not sure is there a way to detect custom elements other than searching for hyphen (`-`) in the tag name. I'm afraid that a simple check `if (nodeName.indexOf('-') !== -1)` may slowdown rendering.

**Comment by zaceno** [comment]
@frenzzy I don't think skipping over children of custom elements is the right approach here. You want to be able to compose the with children just like any other html node. (We want to score well on [custom elements everywhere](https://custom-elements-everywhere.com) eventually )

For example you might want to do:

```
...
<custom-modal titleBarText="blah">
  <Foo state={state.foo}
  <p>{state.bar}</p>
</custom-modal>
```

I haven't really dug into the issue yet, but isn't part of the problem here that the OP isn't using Shadow DOM? If they were, I don't think this would be an issue. So maybe this isn't so much a bug as a question of browser support?

**Comment by zaceno** [comment]
@naltatis I'm not entirely sure why, but I figured out that this works:

```js
  connectedCallback() {
    setTimeout(_ => {
      console.log("example-alert connected - rendering a button");
      this.innerHTML = "<button>alert()</button>";
      this.querySelector("button").addEventListener("click", () => alert("foo"));
    }, 0)
  }
```

**Comment by naltatis** [comment]
Yes using nativ ShadowDOM would solve my issue here, but when you have to support more than the latest browsers this is not an option. You have to use the normal DOM in this case.

As @zaceno pointed out correctly, it’s also not right to generally ignore the children of custom elements. Using ShadowDOM and Slots you need this ability to compose things.

 I’m thinking of an opt in way to tell Hyperapp to skip/ignore the sub tree of a specific node. Maybe as a result of a lifecycle method.

**Comment by naltatis** [comment]
Yes using nativ ShadowDOM would solve my issue here, but when you have to support more than the latest browsers this is not an option. You have to use the normal DOM in this case.

As @zaceno pointed out correctly, it’s also not right to generally ignore the children of custom elements. Using ShadowDOM and Slots you need this ability to compose things.

 I’m thinking of an opt in way to tell Hyperapp to skip/ignore the sub tree of a specific node. Maybe as a result of a lifecycle method.

@zaceno yes the setTimeout would work, but since I don’t control the code inside the custom element this is not an option. I explicitly chose that order of execution to highlight the issue. See the code comment in the first line ;)

**Comment by zaceno** [comment]
@naltatis Ah ok.

You can tell hyperapp to skip the subtree of a specific node if you return the exact same instance of the virtual node, as the previous render -- i e if you memoize the `h('custom-element')` call. 

That only works in the case you're patching though -- not the first render, with hydration, which is what your example is about.

**Comment by zaceno** [comment]
I'm still not exactly sure what's going on. But I verified that this is *only* a problem when "hydrating" -- ie when the pre-rendered markup already contains the `<example-alert>` custom element. If the original markup is clean, the custom element works as intended, and keeps working through continued updates/patching (even without memoization)

Not to say that it's not a problem, but just to point out that if there is a fix for this, it's somewhere in hydration.

**Comment by zaceno** [comment]
Oooh, *now* I get it.

Before hyperapp does it's first render, the markup is this:

```html
<main>
  <h1>click the button</h1>
  <example-alert>
    <button>
      alert()
    </button>
  </example-alert>
</main>
```

... because the custom element inserts the button into itself (*yuck*) when the DOM first mounts. When hyperapp does is about to do the first render, it parses the current DOM -- shown above, including the button -- into a vnode which it compares to the first result of the view function, which is a vnode corresponding to this:

```
<main>
  <h1>click the button</h1>
  <example-alert></example-alert>
</main>
```
-- no button! 

So during the first render, hyperapp does exactly what it should do and *removes* the button (because as far as it knows, the button should no longer be rendered)

Now I also understand why `shouldComponentUpdate` fixes this in React. Our equivalent to `shouldComponentUpdate` is memoization -- but that doesn't work for hydration (because nothing is in the memo yet)

Sorry, I can't see any elegant, simple fixes for this a t m. I guess for now, if you use custom elements you can only have hydration if you also use shadow dom.

**Comment by naltatis** [comment]
> EDIT: It's even more specific than that! It's only a problem if the preexisting markup exactly matches the first render. Any deviation at all (even a whitespace in the markup that wouldn't affect the render) and everything works as expected.
I think I don't quite get what you are saying.

In my example the server rendered markup in the html document exactly matches the markup hyperapp would produce on first render. But the fact, that the custom element cames in earlier and modifies the DOM leads to the fact, that hyperapp removes the unexpected children of the custom element. The workaround tries to fix that by grapping the CEs innerHTML before render and refills it in later (`oncreate`), but looes all events while doing this.

**Comment by zaceno** [comment]
@naltatis yeah, that's expected since you're not moving the actual elements with event listeners, but creating new ones to mimic the previous ones.

**Comment by zaceno** [comment]
Thinking about it a little more. Maybe @frenzzy had a point about children and custom elements though... We shouldn't avoid them during `patch`, but perhaps we should ignore children during hydration (in the `recycleElement` function).

So this: 
```
function recycleElement(element) {
    return {
      nodeName: element.nodeName.toLowerCase(),
      attributes: {},
      children: map.call(element.childNodes, function(element) {
        return element.nodeType === 3 // Node.TEXT_NODE
          ? element.nodeValue
          : recycleElement(element)
      })
    }
  }
```

into this:

```
function recycleElement(element) {
    return {
      nodeName: element.nodeName.toLowerCase(),
      attributes: {},
      children: element.tagName.indexOf('-') ? [] : map.call(element.childNodes, function(element) {
        return element.nodeType === 3 // Node.TEXT_NODE
          ? element.nodeValue
          : recycleElement(element)
      })
    }
  }
```

It still feels a little hacky, but it works (I tried it), and it should prevent custom elements' children (wether static in markup or dynamically generated) from being deleted on first render. On future renders it won't interfere. The only downside here is that we can't have hydration of children of custom elements, even if they were statically declared in the markup.

**Comment by zaceno** [comment]
I know it's not helpful in this case, but I've got to say: my general feeling about using custom-elements in vdom-frameworks is that the only safe bet is to use shadow-dom.  Vdom-engines expect to manage *all* of your dom for you and the more you interfere, the more risk of obscure bugs.

**Comment by naltatis** [comment]
@zaceno I've just tried it in my project. This works perfectly an solves my issue. I had to make a small adjustion and add a `> 0` check to the `indexOf`.

So this is the work code:
```
function recycleElement(element) {
    return {
      nodeName: element.nodeName.toLowerCase(),
      attributes: {},
      children: element.tagName.indexOf('-') > 0 ? [] : map.call(element.childNodes, function(element) {
        return element.nodeType === 3 // Node.TEXT_NODE
          ? element.nodeValue
          : recycleElement(element)
      })
    }
  }
```

To your point on vdom and shadow-dom: Yes I agree, using nativ shadow-dom would be the clean solution and avoid handling cases like this. But in the current situation where some browsers haven't even [started to implement](https://developer.microsoft.com/en-us/microsoft-edge/platform/status/shadowdom/) this necessary to reach that goal eventually.

**Comment by dennisreimann** [comment]
The mentioned workaround using `element.tagName.indexOf('-') > 0` fixes this issue for us as well. Any chance to get this integrated? Thanks to anyone digging deep on the cause of this issue!

**Comment by zaceno** [comment]
@dennisreimann If I were you, I'd make a PR!

The contribution guidelines are here: https://github.com/jorgebucaran/hyperapp/blob/master/CONTRIBUTING.md

Beyond the info in there, It's also a good idea to prepare a small example in codepen or jsfiddle which demonstrates the problem being solved, and also to do some thinking about what (if anything) might be affected by this change. Does it have any downsides? Any ways it could be improved?

**Comment by TimOetting** [comment]
I am in the process of migrating the exact project that @naltatis has been working on to V2. Is there any solution in the new version to leave custom elements unchanged during hydration?

I tried to adapt the change being made in the `recycleElement` function to V2 but the custom elements` content still gets removed.

```
var recycleNode = (node) => {
  return node.nodeType === TEXT_NODE
    ? text(node.nodeValue, node)
    : createVNode(
        node.nodeName.toLowerCase(),
        EMPTY_OBJ,
        node.tagName.indexOf('-') > 0 ? [] : map.call(node.childNodes, recycleNode),
        SSR_NODE,
        node
      )}
```

As @jorgebucaran mentioned in #764 , the DOM mounting mechanism has been changed in V2 and mentions moving the hydration code to an external package. However, the issue is already 4 years old. I still have to sort my head out on how an external hydration package would fit in here. Would this still be the way to go for you?

**Comment by TimOetting** [comment]
I just realized that a misunderstanding on my part caused the app to re-render my entire root node during hydration and my above attempts failed because of it. Unlike V1, Hyperapp now seems to replace the initial node with the view instead of using it as the parent node. Having this fixed, the code works so far.

I am now looking again at @naltati's PR and see if I can derive a PR for V2 from that. Currently I'm obviously still missing the `isRecycling` variable in the right place, but that will surely be possible somehow.

**Comment by kofifus** [comment]
@TimOetting Hi can you please share the code you mentions works above in V2 ?
